org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/filesystem/JLSFile.java

```diff
@@ -0,0 +1,105 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.filesystem;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import org.eclipse.core.filesystem.IFileStore;
+import org.eclipse.core.internal.filesystem.local.LocalFile;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+
+/**
+ * JDT.LS's own implementation of files in the local operating system's file system.
+ * The instance of this class will be returned by {@link JLSFileSystem}.
+ */
+public class JLSFile extends LocalFile {
+
+    public JLSFile(File file) {
+        super(file);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * <p>
+     * Since the metadata files may be redirected into the workspace,
+     * we override the method to make sure those files are not missed.
+     * </p>
+     */
+    @Override
+    public String[] childNames(int options, IProgressMonitor monitor) {
+        String[] childNames = super.childNames(options, monitor);
+        if (JLSFsUtils.generatesMetadataFilesAtProjectRoot()) {
+            return childNames;
+        }
+
+        IPath filePath = ResourceUtils.filePathFromURI(this.toURI().toString());
```

Comment:
You use multiple different approaches to get the IPath object in `childNames`, `getChild` and `getFileStore` methods.
- IPath filePath = ResourceUtils.filePathFromURI(this.toURI().toString());
- IPath path = new Path(this.filePath).append(name);
- IPath fullPath = new Path(file.getPath()).append(path);

 I think the faster way is using `new Path(this.filePath)` directly.